packages/components/src/composite/legacy.tsx

```diff
@@ -0,0 +1,174 @@
+	Parameters< typeof Current.useCompositeStore >[ 0 ]
+>;
+
+type Orientation = 'horizontal' | 'vertical';
+
+export interface InitialState {
+	baseId?: string;
+	unstable_virtual?: boolean;
+	rtl?: boolean;
+	orientation?: Orientation;
+	currentId?: string;
+	loop?: boolean | Orientation;
+	wrap?: boolean | Orientation;
+	shift?: boolean;
+}
+
+type BaseIdState = Required< Pick< InitialState, 'baseId' > >;
+type BaseId = BaseIdState[ 'baseId' ];
+type MappedInitialState = CompositeStoreProps & BaseIdState;
+
+export type CompositeState = CompositeStoreState & BaseIdState;
+export type CompositeStateProps =
+	| { state: CompositeState }
+	| ( CompositeState & { state?: never } );
+export type CompositeProps< C extends Component > = CompositeStateProps &
+	Parameters< C >[ 0 ];
+
+type ManagedProps = CompositeStoreState & {
+	props: Record< PropertyKey, any >;
+};
+
+function showDeprecationMessage( ...props: Parameters< typeof deprecated > ) {
+	if ( 'test' !== process.env.NODE_ENV ) {
+		deprecated( ...props );
+	}
+}
+
+const idMap = new Map< BaseId, React.MutableRefObject< number > >();
+
+function useBaseId( preferredBaseId?: BaseId ): BaseId {
+	const generatedId = useGeneratedId();
+	const [ baseId ] = useState( preferredBaseId || generatedId );
+	idMap.set( baseId, useRef( 0 ) );
+
+	return baseId;
+}
+
+function useId( baseId: BaseId, preferredId?: string ) {
+	const counter = idMap.get( baseId ) ?? { current: 0 };
+	const [ id ] = useState( () => {
+		if ( preferredId ) return preferredId;
+		return `${ baseId }-${ ++counter.current }`;
+	} );
+
+	return id;
+}
+
+function useMappedInitialState(
+	initialState: InitialState = {}
+): MappedInitialState {
+	const {
+		baseId,
+		currentId: defaultActiveId,
+		orientation,
+		rtl = false,
+		loop: focusLoop = false,
+		wrap: focusWrap = false,
+		shift: focusShift = false,
+		// eslint-disable-next-line camelcase
+		unstable_virtual: virtualFocus,
+	} = initialState;
+
+	return {
+		baseId: useBaseId( baseId ),
+		defaultActiveId,
+		rtl,
+		orientation,
+		focusLoop,
+		focusShift,
+		focusWrap,
+		virtualFocus,
+	};
+}
+
+function manageProps( stateProps: CompositeStateProps ): ManagedProps {
+	if ( stateProps.state ) {
+		const { state, ...rest } = stateProps;
+		const { store, props } = manageProps( state );
+		return { store, props: { ...rest, ...props } };
+	}
+
+	const { store, ...props } = stateProps;
+	return { store, props };
+}
+
+function proxyComposite< C extends Component >(
+	LegacyComponent: C | ( ( ...args: any[] ) => C ),
+	propMap: Record< PropertyKey, PropertyKey | null > = {}
+): ( props: CompositeProps< C > ) => React.ReactElement {
+	return ( unmanagedProps ) => {
+		const componentName =
+			// @ts-ignore
+			LegacyComponent.displayName ?? LegacyComponent.render.name;
+
+		showDeprecationMessage( componentName, {
+			alternative: `@wordpress/components:${ componentName }`,
+		} );
+
+		const { store, props } = manageProps( unmanagedProps );
```

Comment:
Nit: could we rename some of these variables / functions to clarify better what is the intent of the code? Maybe something like

```suggestion
		const { store, props } = mapLegacyPropsToCurrentProps( legacyProps );
```